---
title: "Founders Cocaine vs Sham Expression Analysis"
author: "Michael C. Saul (michael.saul [at] jax.org)"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
    number_sections: true
    fig_retina: true
    self_contained: true
    code_folding: show
    theme: flatly
    highlight: default
    df_print: paged
---

# Background

## Motivation

### Purpose of analysis

This analysis is performed to model striatal genes whose expression is associated with strain-specific correlates of cocaine exposure.

Also assuming the possibility of interesting interactions between cocaine and sex on gene expression.

# Analysis

## Setup

### Working directory

All chunks will run in the working folder for this experiment.

```{r setup}
knitr::opts_knit$set(root.dir = "~/Box/projects/founders_rnaseq/")
```

### R libraries

Calling R libraries necessary for this analysis.

```{r}
library("limma")
library("edgeR")
library("ggplot2")
library("ggbeeswarm")
library("gplots")
library("RColorBrewer")
library("biomaRt")
library("statmod")
```

### Importing data

Importing data generated by `founders_expression_prep_STAR_RSEM.Rmd`.

```{r}
founders_DGEList = readRDS(file = "~/Box/projects/founders_rnaseq/expression_study/founders_128_RSEM_DGEList_2019-08-19.RDS")
founders_voom    = readRDS(file = "~/Box/projects/founders_rnaseq/expression_study/founders_128_RSEM_voom_2019-08-19.RDS")
founders_pheno   = readRDS("~/Box/projects/founders_rnaseq/expression_study/founders_128_RSEM_pheno_2019-08-19.RDS")
exprs_pheno_equal = length(which(row.names(founders_pheno) == colnames(founders_voom))) == ncol(founders_voom)
ifelse(exprs_pheno_equal, "Expression and phenotype data have the same sample IDs.", "Mismatch between expression and phenotype sample IDs.")
```

## Gene Expression Analysis

### Sham expression heritability

The heritability of sham-exponsed gene expression in the founders is examined here. We use this equation:

$h^2=\displaystyle{{MS_{strain}-MS_{resid}}\over{MS_{strain} + (n_{mean} - 1) * MS_{resid}}}$

```{r}
samples_sham = founders_pheno[which(founders_pheno$Injection == "Sham"),"Name"]
strain_sham = founders_pheno[samples_sham,"Strain"]
sex_sham = founders_pheno[samples_sham,"chrY_sex"]
choroid_plexus_sham = founders_pheno[samples_sham,"choroid_plexus_covariate"]

for (i in 1:nrow(founders_voom)) {
  gene_i = row.names(founders_voom)[i]
  sham_expr_i = founders_voom$E[i,samples_sham]
  lm_i = lm(sham_expr_i ~ choroid_plexus_sham + sex_sham + strain_sham + sex_sham:strain_sham)
  aic_i = AIC(lm_i)
  anova_i = anova(lm_i)
  lm_nosex_i = lm(sham_expr_i ~ choroid_plexus_sham + strain_sham)
  aic_nosex_i = AIC(lm_nosex_i)
  anova_nosex_i = anova(lm_nosex_i)
  sham_heritability_i = (anova_i$`Mean Sq`[3] - anova_i$`Mean Sq`[4]) / (2 * (anova_i$`Mean Sq`[3] + (8 - 1) * anova_i$`Mean Sq`[5]))
  sham_heritability_nosex_i = (anova_nosex_i$`Mean Sq`[2] - anova_nosex_i$`Mean Sq`[3]) / (anova_nosex_i$`Mean Sq`[2] + ((8 - 1) * anova_nosex_i$`Mean Sq`[3]))
  df_i = data.frame(gene = gene_i, h2_sex_inter = sham_heritability_i, h2_nosex = sham_heritability_nosex_i,
                    aic_sex_inter = aic_i, aic_nosex = aic_nosex_i, stringsAsFactors = FALSE)
  if (exists("h2_df")) {
    h2_df = rbind(h2_df, df_i)
  } else {
    h2_df = df_i
  }
}

h2_df[which(h2_df$h2_sex_inter <= 0),"h2_sex_inter"] = 0
h2_df[which(h2_df$h2_nosex <= 0),"h2_nosex"] = 0
row.names(h2_df) = h2_df$gene
h2_df$h2_lowest_aic = ifelse(h2_df$aic_nosex < h2_df$aic_sex_inter, h2_df$h2_nosex, h2_df$h2_sex_inter)

hist(h2_df$h2_lowest_aic, breaks = 50, col = "#8A1B61", xlab = "Sham Expression Heritability", main = "Heritability of Founder Striatum Expression", xlim = c(0,1))
saveRDS(h2_df, paste("founders_sham_expression_heritabilities_",format(Sys.time(),"%Y-%m-%d"),".RDS",sep=""))
```

Tabulating median no-sex heritability

```{r median_nosex_heritability}
# R code

median(h2_df$h2_nosex)
```

### Cocaine vs. sham factorial experiment: setup

Getting phenotypes of interest.

```{r}
# Model factors for founders model
founders_strain = gsub("/","",founders_pheno$Strain)
founders_sex = founders_pheno$chrY_sex
founders_cocaine = founders_pheno$Injection
founders_choroidplexus = founders_pheno$choroid_plexus_covariate
founders_log2_choroidplexus = log2(founders_choroidplexus)
```

Making strain-specific toptables.

```{r}
strains = unique(founders_strain)
for (i in strains) {
  
  samples_i = which(founders_strain == i)
  sex_i = founders_sex[samples_i]
  cocaine_i = founders_cocaine[samples_i]
  choroidplexus_i = founders_log2_choroidplexus[samples_i]
  voom_i = founders_voom[,samples_i]
  modmat_i = model.matrix(samples_i ~ choroidplexus_i + sex_i + cocaine_i + sex_i:cocaine_i)
  
  lmFit_i = lmFit(voom_i, design = modmat_i, method = "robust", weights = NULL)
  lmFit_i = eBayes(lmFit_i, trend = TRUE, robust = TRUE)
  
  toptable_sex_i = as.data.frame(topTable(lmFit_i,
                                          coef = grep("^sex_i[MF]$", colnames(modmat_i)),
                                          number = nrow(voom_i)))
  toptable_cocaine_i = as.data.frame(topTable(lmFit_i,
                                              coef = grep("^cocaine_i\\w{3,7}$", colnames(modmat_i)),
                                              number = nrow(voom_i)))
  toptable_interact_i = as.data.frame(topTable(lmFit_i,
                                               coef = grep("^sex_i[MF]:cocaine_i\\w{3,7}$", colnames(modmat_i)),
                                               number = nrow(voom_i)))
  
  assign(paste("strain",i,"sex_toptable",sep="_"), toptable_sex_i)
  assign(paste("strain",i,"cocaine_toptable",sep="_"), toptable_cocaine_i)
  assign(paste("strain",i,"interaction_toptable",sep="_"), toptable_interact_i)
  
  cat(length(which(toptable_sex_i$adj.P.Val < 0.01)), " genes in strain ", i, " differentially expressed across sexes at FDR < 0.01.\n", sep = "")
  cat(length(which(toptable_sex_i$adj.P.Val < 0.05)), " genes in strain ", i, " differentially expressed across sexes at FDR < 0.05.\n", sep = "")
  cat(length(which(toptable_sex_i$adj.P.Val < 0.10)), " genes in strain ", i, " differentially expressed across sexes at FDR < 0.10.\n", sep = "")
  cat(length(which(toptable_sex_i$adj.P.Val < 0.20)), " genes in strain ", i, " differentially expressed across sexes at FDR < 0.20.\n\n", sep = "")
  
  cat(length(which(toptable_cocaine_i$adj.P.Val < 0.01)), " genes in strain ", i, " differentially expressed due to cocaine at FDR < 0.01.\n", sep = "")
  cat(length(which(toptable_cocaine_i$adj.P.Val < 0.05)), " genes in strain ", i, " differentially expressed due to cocaine at FDR < 0.05.\n", sep = "")
  cat(length(which(toptable_cocaine_i$adj.P.Val < 0.10)), " genes in strain ", i, " differentially expressed due to cocaine at FDR < 0.10.\n", sep = "")
  cat(length(which(toptable_cocaine_i$adj.P.Val < 0.20)), " genes in strain ", i, " differentially expressed due to cocaine at FDR < 0.20.\n\n", sep = "")
  
  cat(length(which(toptable_interact_i$adj.P.Val < 0.01)), " genes in strain ", i, " differentially expressed in sex:cocaine interaction at FDR < 0.01.\n", sep = "")
  cat(length(which(toptable_interact_i$adj.P.Val < 0.05)), " genes in strain ", i, " differentially expressed in sex:cocaine interaction at FDR < 0.05.\n", sep = "")
  cat(length(which(toptable_interact_i$adj.P.Val < 0.10)), " genes in strain ", i, " differentially expressed in sex:cocaine interaction at FDR < 0.10.\n", sep = "")
  cat(length(which(toptable_interact_i$adj.P.Val < 0.20)), " genes in strain ", i, " differentially expressed in sex:cocaine interaction at FDR < 0.20.\n\n\n", sep = "")
  
  rm(list = ls()[grep("_i$",ls())])
}
```

Constructing overall model matrix.

```{r}
# Model matrix for founders factorial experiment with choroid plexus run as an additive covariate
founders_modmat = model.matrix(~ founders_log2_choroidplexus +
                                 founders_strain *
                                 founders_sex *
                                 founders_cocaine)
colnames(founders_modmat) = gsub("founders_","",colnames(founders_modmat))
colnames(founders_modmat)
```

Using voom+limma with robust variance estimation. Fitting the model.

```{r}
founders_lmFit = lmFit(founders_voom, design = founders_modmat, method = "robust", weights = NULL)
founders_lmFit = eBayes(founders_lmFit, trend = TRUE, robust = TRUE)
```

Summarizing the genes differentially expressed across strains.

```{r}
founders_strain_toptable = as.data.frame(topTable(founders_lmFit, 
                                                  coef = grep("^strain\\w{2,4}$", colnames(founders_modmat)), 
                                                  number = nrow(founders_voom)))
paste(length(which(founders_strain_toptable$adj.P.Val < 0.01)), " genes differentially expressed across strains at FDR < 0.01.", sep = "")
paste(length(which(founders_strain_toptable$adj.P.Val < 0.05)), " genes differentially expressed across strains at FDR < 0.05.", sep = "")
paste(length(which(founders_strain_toptable$adj.P.Val < 0.10)), " genes differentially expressed across strains at FDR < 0.10.", sep = "")
paste(length(which(founders_strain_toptable$adj.P.Val < 0.20)), " genes differentially expressed across strains at FDR < 0.20.", sep = "")
```

Summarizing the genes differentially expressed across sexes.

```{r}
founders_sex_toptable = as.data.frame(topTable(founders_lmFit, 
                                               coef = grep("^sexM$", colnames(founders_modmat)), 
                                               number = nrow(founders_voom)))
paste(length(which(founders_sex_toptable$adj.P.Val < 0.01)), " genes differentially expressed across sexes at FDR < 0.01.", sep = "")
paste(length(which(founders_sex_toptable$adj.P.Val < 0.05)), " genes differentially expressed across sexes at FDR < 0.05.", sep = "")
paste(length(which(founders_sex_toptable$adj.P.Val < 0.10)), " genes differentially expressed across sexes at FDR < 0.10.", sep = "")
paste(length(which(founders_sex_toptable$adj.P.Val < 0.20)), " genes differentially expressed across sexes at FDR < 0.20.", sep = "")
```

Summarizing the genes differentially expressed between cocaine and sham.

```{r}
founders_cocaine_toptable = as.data.frame(topTable(founders_lmFit, 
                                                   coef = grep("^cocaineSham$", colnames(founders_modmat)), 
                                                   number = nrow(founders_voom)))
paste(length(which(founders_cocaine_toptable$adj.P.Val < 0.01)), " genes differentially expressed between cocaine and sham at FDR < 0.01.", sep = "")
paste(length(which(founders_cocaine_toptable$adj.P.Val < 0.05)), " genes differentially expressed between cocaine and sham at FDR < 0.05.", sep = "")
paste(length(which(founders_cocaine_toptable$adj.P.Val < 0.10)), " genes differentially expressed between cocaine and sham at FDR < 0.10.", sep = "")
paste(length(which(founders_cocaine_toptable$adj.P.Val < 0.20)), " genes differentially expressed between cocaine and sham at FDR < 0.20.", sep = "")
```

Summarizing the genes interacting between strain and sex.

```{r}
founders_strain_sex_toptable = as.data.frame(topTable(founders_lmFit,
                                                      coef = grep("^strain\\w{2,4}:sexM$", colnames(founders_modmat)),
                                                      number = nrow(founders_voom)))
paste(length(which(founders_strain_sex_toptable$adj.P.Val < 0.01)), " genes whose expression depends upon both strain and sex at FDR < 0.01.", sep = "")
paste(length(which(founders_strain_sex_toptable$adj.P.Val < 0.05)), " genes whose expression depends upon both strain and sex at FDR < 0.05.", sep = "")
paste(length(which(founders_strain_sex_toptable$adj.P.Val < 0.10)), " genes whose expression depends upon both strain and sex at FDR < 0.10.", sep = "")
paste(length(which(founders_strain_sex_toptable$adj.P.Val < 0.20)), " genes whose expression depends upon both strain and sex at FDR < 0.20.", sep = "")
```

Summarizing the genes interacting between strains and cocaine exposure.

```{r}
founders_strain_cocaine_toptable = as.data.frame(topTable(founders_lmFit,
                                                          coef = grep("^strain\\w{2,4}:cocaineSham$", colnames(founders_modmat)),
                                                          number = nrow(founders_voom)))
paste(length(which(founders_strain_cocaine_toptable$adj.P.Val < 0.01)), " genes whose expression depends upon both cocaine exposure and strain at FDR < 0.01.", sep = "")
paste(length(which(founders_strain_cocaine_toptable$adj.P.Val < 0.05)), " genes whose expression depends upon both cocaine exposure and strain at FDR < 0.05.", sep = "")
paste(length(which(founders_strain_cocaine_toptable$adj.P.Val < 0.10)), " genes whose expression depends upon both cocaine exposure and strain at FDR < 0.10.", sep = "")
paste(length(which(founders_strain_cocaine_toptable$adj.P.Val < 0.20)), " genes whose expression depends upon both cocaine exposure and strain at FDR < 0.20.", sep = "")
```

Summarizing the genes interacting between sex and cocaine exposure.

```{r}
founders_sex_cocaine_toptable = as.data.frame(topTable(founders_lmFit,
                                                       coef = grep("^sexM:cocaineSham$", colnames(founders_modmat)),
                                                       number = nrow(founders_voom)))
paste(length(which(founders_sex_cocaine_toptable$adj.P.Val < 0.01)), " genes whose expression depends upon both cocaine exposure and sex at FDR < 0.01.", sep = "")
paste(length(which(founders_sex_cocaine_toptable$adj.P.Val < 0.05)), " genes whose expression depends upon both cocaine exposure and sex at FDR < 0.05.", sep = "")
paste(length(which(founders_sex_cocaine_toptable$adj.P.Val < 0.10)), " genes whose expression depends upon both cocaine exposure and sex at FDR < 0.10.", sep = "")
paste(length(which(founders_sex_cocaine_toptable$adj.P.Val < 0.20)), " genes whose expression depends upon both cocaine exposure and sex at FDR < 0.20.", sep = "")
```

Finally, looking at strain:sex:cocaine interaction

```{r}
founders_strain_sex_cocaine_toptable = as.data.frame(topTable(founders_lmFit,
                                                              coef = grep("^strain\\w{2,4}:sexM:cocaineSham$", colnames(founders_modmat)),
                                                              number = nrow(founders_voom)))
paste(length(which(founders_strain_sex_cocaine_toptable$adj.P.Val < 0.01)), " genes whose expression depends upon cocaine exposure, sex, and strain at FDR < 0.01.", sep = "")
paste(length(which(founders_strain_sex_cocaine_toptable$adj.P.Val < 0.05)), " genes whose expression depends upon cocaine exposure, sex, and strain at FDR < 0.05.", sep = "")
paste(length(which(founders_strain_sex_cocaine_toptable$adj.P.Val < 0.10)), " genes whose expression depends upon cocaine exposure, sex, and strain at FDR < 0.10.", sep = "")
paste(length(which(founders_strain_sex_cocaine_toptable$adj.P.Val < 0.20)), " genes whose expression depends upon cocaine exposure, sex, and strain at FDR < 0.20.", sep = "")
```

### Plotting interesting genes

Making plots: building plotting data frame.

```{r}
strain_key = readRDS("~/Box/CC_DO_founders/cc_do_founders_key.RDS")
founders_voom_df = cbind(t(founders_voom$E), founders_pheno)
founders_voom_df$Strain = gsub("/","",founders_voom_df$Strain)
founders_voom_df$Strain = factor(founders_voom_df$Strain,
                                 levels = strain_key$collaborative_cross_abbreviation,
                                 ordered = TRUE)
founders_voom_df$Injection = factor(founders_voom_df$Injection,
                                    c("Sham","Cocaine"),
                                    ordered = TRUE)
founders_voom_df$Sex = founders_voom_df$chrY_Sex

strain_sex_cocaine_plot = function(ensembl_id, gene_id = ensembl_id) {
  gene_formula = paste(ensembl_id," ~ Strain + chrY_sex + Injection", sep = "")
  
  gene = aggregate(eval(parse(text = gene_formula)),
                   data = founders_voom_df, FUN = mean)
  colnames(gene)[4] = "gene_mean"
  gene_sd = aggregate(eval(parse(text = gene_formula)), 
                      data = founders_voom_df, FUN = sd)
  gene_length = aggregate(eval(parse(text = gene_formula)),
                          data = founders_voom_df, FUN = length)
  gene$gene_sd = gene_sd[,4]
  gene$gene_length = gene_length[,4]
  gene$gene_sem = gene$gene_sd / sqrt(gene$gene_length)
  gene$gene_ul = gene$gene_mean + gene$gene_sem
  gene$gene_ll = gene$gene_mean - gene$gene_sem
  gene$Strain = factor(gene$Strain,
                       levels = strain_key$collaborative_cross_abbreviation,
                       ordered = TRUE)
  gene$Injection = factor(gene$Injection,
                          levels = c("Sham","Cocaine"),
                          ordered = TRUE)
  gene_plot = ggplot(data = gene, aes(x = Strain, y = gene_mean, color = Strain, shape = Injection)) +
    geom_errorbar(aes(ymin = gene_ll, ymax = gene_ul), width = 0.6, position = position_dodge(width = 0.75), size = 1) +
    geom_point(size = 3, position = position_dodge(width = 0.75), fill = "white") +
    geom_beeswarm(data = founders_voom_df, 
                  aes(x = Strain, y = eval(parse(text = ensembl_id)), color = Strain, shape = Injection),
                  alpha = 0.5, dodge.width = 0.75, size = 1, fill = "white") +
    theme_bw() +
    xlab("Founder Strain") +
    ylab(paste(gene_id," expression log2(CPM)", sep = "")) +
    facet_grid(chrY_sex ~ ., scales = "free_y") +
    scale_color_manual(values = strain_key$collaborative_cross_color_broman) +
    scale_shape_manual(values = c(23,16)) + 
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
  return(gene_plot)
}

strain_cocaine_plot = function(ensembl_id, gene_id = ensembl_id) {
  gene_formula = paste(ensembl_id," ~ Strain + Injection", sep = "")
  gene = aggregate(eval(parse(text = gene_formula)),
                   data = founders_voom_df, FUN = mean)
  colnames(gene)[3] = "gene_mean"
  gene_sd = aggregate(eval(parse(text = gene_formula)), 
                      data = founders_voom_df, FUN = sd)
  gene_length = aggregate(eval(parse(text = gene_formula)),
                          data = founders_voom_df, FUN = length)
  gene$gene_sd = gene_sd[,3]
  gene$gene_length = gene_length[,3]
  gene$gene_sem = gene$gene_sd / sqrt(gene$gene_length)
  gene$gene_ul = gene$gene_mean + gene$gene_sem
  gene$gene_ll = gene$gene_mean - gene$gene_sem
  gene$Strain = factor(gene$Strain,
                       levels = strain_key$collaborative_cross_abbreviation,
                       ordered = TRUE)
  gene$Injection = factor(gene$Injection,
                          levels = c("Sham","Cocaine"),
                          ordered = TRUE)
  gene_plot = ggplot(data = gene, aes(x = Strain, y = gene_mean, color = Strain, shape = Injection)) +
    geom_errorbar(aes(ymin = gene_ll, ymax = gene_ul), width = 0.6, position = position_dodge(width = 0.75), size = 1) +
    geom_point(size = 3, position = position_dodge(width = 0.75), fill = "white") +
    geom_beeswarm(data = founders_voom_df, 
                  aes(x = Strain, y = eval(parse(text = ensembl_id)), color = Strain, shape = Injection),
                  alpha = 0.5, dodge.width = 0.75, size = 1, fill = "white") +
    theme_bw() +
    xlab("Founder Strain") +
    ylab(paste(gene_id," expression log2(CPM)", sep = "")) +
    scale_color_manual(values = strain_key$collaborative_cross_color_broman) +
    scale_shape_manual(values = c(23,16)) + 
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
  return(gene_plot)
}

strain_sex_cocaine_beeswarm = function(ensembl_id, gene_id = ensembl_id) {
  gene_formula = paste(ensembl_id," ~ Strain + Sex + Injection", sep = "")
  gene = aggregate(eval(parse(text = gene_formula)),
                   data = founders_voom_df, FUN = mean)
  colnames(gene)[4] = "gene_mean"
  gene_sd = aggregate(eval(parse(text = gene_formula)), 
                      data = founders_voom_df, FUN = sd)
  gene_length = aggregate(eval(parse(text = gene_formula)),
                          data = founders_voom_df, FUN = sd)
  gene$gene_sd = gene_sd[,4]
  gene$gene_length = gene_length[,4]
  gene$gene_sem = gene$gene_sd / sqrt(gene$gene_length)
  gene$gene_ul = gene$gene_mean + gene$gene_sem
  gene$gene_ll = gene$gene_mean - gene$gene_sem
  gene$Strain = factor(gene$Strain,
                       levels = strain_key$collaborative_cross_abbreviation,
                       ordered = TRUE)
  gene$Injection = factor(gene$Injection,
                          levels = c("Sham","Cocaine"),
                          ordered = TRUE)
  gene_plot = ggplot(data = gene, aes(x = Strain, y = gene_mean, color = Strain, shape = Injection)) +
    geom_beeswarm(data = founders_voom_df, 
                  aes(x = Strain, y = eval(parse(text = ensembl_id)), color = Strain, shape = Injection),
                  dodge.width = 0.75, size = 3, fill = "white") +
    theme_bw() +
    xlab("Founder Strain") +
    ylab(paste(gene_id," expression log2(CPM)", sep = "")) +
    facet_grid(Sex ~ ., scales = "free_y") +
    scale_color_manual(values = strain_key$collaborative_cross_color_broman) +
    scale_shape_manual(values = c(23,16)) + 
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
  return(gene_plot)
}
```

### Saving Data

Saving all data

```{r}
save(list = ls(), file = paste("founders_RSEM_expression_plots_data_",format(Sys.time(),"%Y-%m-%d"),".Rdata",sep=""))
```

## Reproducibility Information

### R session and OS information

```{r sessionInfo}
founders_expression_sessionInfo = sessionInfo()
founders_expression_sessionInfo
saveRDS(founders_expression_sessionInfo,paste("./founders_expression_sessionInfo_",format(Sys.time(),"%Y-%m-%d"),"_sessionInfo.RDS",sep=""))
```

### Document Control

This document was prepared using RMarkdown in RStudio.