---
title: "Founders Cocaine vs Sham: Expression Overlap with Human Analysis"
author: "Michael C. Saul (michael.saul [at] jax.org)"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
    number_sections: true
    fig_retina: true
    self_contained: true
    code_folding: show
    theme: flatly
    highlight: default
    df_print: paged
---

# Background

## Motivation

### Purpose of analysis

This analysis is performed to identify overlapping striatum cocaine-related gene sets whose expression overlaps with human-specific gene expression correlates of cocaine use disorder..

# Analysis

## Setup

### Working directory

All chunks will run in the working folder for this experiment.

```{r setup}
knitr::opts_knit$set(root.dir = "~/Box/projects/founders_rnaseq/")
```

### R libraries

Calling R libraries necessary for this analysis.

```{r}
library("limma")
library("edgeR")
library("ggplot2")
library("ggbeeswarm")
library("ggrepel")
library("gplots")
library("RColorBrewer")
library("biomaRt")
library("statmod")
library("qvalue")
library("msaul")
library("RCurl")
library("biomaRt")
library("tidyverse")
library("viridis")
```

### Importing Expression Data

Importing data generated by `founders_expression_prep_STAR_RSEM.Rmd`.

```{r}
founders_DGEList = readRDS(file = "~/Box/projects/founders_rnaseq/expression_study/founders_128_RSEM_DGEList_2019-08-19.RDS")
founders_voom    = readRDS(file = "~/Box/projects/founders_rnaseq/expression_study/founders_128_RSEM_voom_2019-08-19.RDS")
founders_pheno   = readRDS("~/Box/projects/founders_rnaseq/expression_study/founders_128_RSEM_pheno_2019-08-19.RDS")
exprs_pheno_equal = length(which(row.names(founders_pheno) == colnames(founders_voom))) == ncol(founders_voom)
ifelse(exprs_pheno_equal, "Expression and phenotype data have the same sample IDs.", "Mismatch between expression and phenotype sample IDs.")
```

### Getting mouse-human homology

Data accessed `r format(Sys.time(),"%Y-%m-%d")`.

```{r get_homology}
# Downloading homology report from MGI
Mm_Hs_homology_URL = getURL("http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt")
Mm_Hs_homology_raw = read.table(text = Mm_Hs_homology_URL,
                                sep = "\t",
                                header = TRUE,
                                stringsAsFactors = FALSE)

# Removing homology report URL to keep environment clean
rm(list = c("Mm_Hs_homology_URL"))

# Filtering to identify 1:1 orthologs and summarize into a useful data frame.
Mm_Hs_homology_1to1 = Mm_Hs_homology_raw %>%
  group_by(HomoloGene.ID) %>%
  mutate(n_human_genes = length(which(NCBI.Taxon.ID == 9606)),
         n_mouse_genes = length(which(NCBI.Taxon.ID == 10090))) %>%
  filter(n_human_genes == 1 & n_mouse_genes == 1) %>%
  summarize(Hs_Symbol = Symbol[which(NCBI.Taxon.ID == 9606)],
            Mm_Symbol = Symbol[which(NCBI.Taxon.ID == 10090)],
            Hs_EntrezGene_ID = EntrezGene.ID[which(NCBI.Taxon.ID == 9606)],
            Mm_EntrezGene_ID = EntrezGene.ID[which(NCBI.Taxon.ID == 10090)],
            Hs_HGNC_ID = HGNC.ID[which(NCBI.Taxon.ID == 9606)],
            HS_OMIM_ID = OMIM.Gene.ID[which(NCBI.Taxon.ID == 9606)],
            Mm_MGI_ID = Mouse.MGI.ID[which(NCBI.Taxon.ID == 10090)],
            Hs_RefSeq_IDs = Nucleotide.RefSeq.IDs[which(NCBI.Taxon.ID == 9606)],
            Mm_RefSeq_IDs = Nucleotide.RefSeq.IDs[which(NCBI.Taxon.ID == 10090)],
            Hs_Genomic_Coordinates = Genomic.Coordinates..mouse....human...[which(NCBI.Taxon.ID == 9606)],
            Mm_Genomic_Coordinates = Genomic.Coordinates..mouse....human...[which(NCBI.Taxon.ID == 10090)]) %>%
  mutate(Hs_genomic_chr = gsub("^Chr(.+):\\d+-\\d+\\([-+]\\)$","\\1",Hs_Genomic_Coordinates),
         Hs_genomic_start = as.numeric(gsub("^Chr.+:(\\d+)-\\d+\\([-+]\\)$","\\1",Hs_Genomic_Coordinates)),
         Hs_genomic_stop = as.numeric(gsub("^Chr.+:\\d+-(\\d+)\\([-+]\\)$","\\1",Hs_Genomic_Coordinates)),
         Hs_genomic_strand = gsub("^Chr.+:\\d+-\\d+\\(([-+])\\)$","\\1",Hs_Genomic_Coordinates),
         Mm_genomic_chr = gsub("^Chr(.+):\\d+-\\d+\\([-+]\\)$","\\1",Mm_Genomic_Coordinates),
         Mm_genomic_start = as.numeric(gsub("^Chr.+:(\\d+)-\\d+\\([-+]\\)$","\\1",Mm_Genomic_Coordinates)),
         Mm_genomic_stop = as.numeric(gsub("^Chr.+:\\d+-(\\d+)\\([-+]\\)$","\\1",Mm_Genomic_Coordinates)),
         Mm_genomic_strand = gsub("^Chr.+:\\d+-\\d+\\(([-+])\\)$","\\1",Mm_Genomic_Coordinates)) %>%
  dplyr::select(-Hs_Genomic_Coordinates, -Mm_Genomic_Coordinates)

# Changing to a data frame and fixing the HomoloGene ID column
Mm_Hs_homology_1to1 = as.data.frame(Mm_Hs_homology_1to1)
colnames(Mm_Hs_homology_1to1)[1] = "HomoloGene_ID"

# Ensuring that the gene symbols act as good and unique database IDs
cat(ifelse(length(unique(Mm_Hs_homology_1to1$Hs_Symbol)) == nrow(Mm_Hs_homology_1to1) & length(unique(Mm_Hs_homology_1to1$Mm_Symbol)) == nrow(Mm_Hs_homology_1to1),
           "All human and mouse gene symbols are unique.", "Some duplication in mouse or human gene symbols."))
```

## Gene Expression Analysis

### Cocaine vs. sham factorial experiment: setup

Getting phenotypes of interest.

```{r}
# Model factors for founders model
founders_strain = gsub("/","",founders_pheno$Strain)
founders_sex = founders_pheno$Sex
founders_cocaine = founders_pheno$Injection
founders_choroidplexus = founders_pheno$choroid_plexus_covariate
founders_log2_choroidplexus = log2(founders_choroidplexus)
```

Making strain-specific toptables.

```{r warning=FALSE}
founders_strain_sex_cocaine = paste(founders_strain, founders_sex, founders_cocaine, sep = "_")
modmat_contrasts = model.matrix(~ 0 + founders_strain_sex_cocaine + founders_log2_choroidplexus)
# modmat_edgeR_contrasts = model.matrix(~ 0 + founders_strain_sex_cocaine + founders_choroidplexus)
colnames(modmat_contrasts) = gsub("^founders_strain_sex_cocaine","s",colnames(modmat_contrasts))
founders_cocaine_overall_contrasts = makeContrasts(all_cocaine = (s129_F_Cocaine - s129_F_Sham) +
                                                     (sAJ_F_Cocaine - sAJ_F_Sham) +
                                                     (sB6_F_Cocaine - sB6_F_Sham) +
                                                     (sNOD_F_Cocaine - sNOD_F_Sham) +
                                                     (sNZO_F_Cocaine - sNZO_F_Sham) +
                                                     (sCAST_F_Cocaine - sCAST_F_Sham) +
                                                     (sPWK_F_Cocaine - sPWK_F_Sham) +
                                                     (sWSB_F_Cocaine - sWSB_F_Sham) +
                                                     (s129_M_Cocaine - s129_M_Sham) +
                                                     (sAJ_M_Cocaine - sAJ_M_Sham) +
                                                     (sB6_M_Cocaine - sB6_M_Sham) +
                                                     (sNOD_M_Cocaine - sNOD_M_Sham) +
                                                     (sNZO_M_Cocaine - sNZO_M_Sham) +
                                                     (sCAST_M_Cocaine - sCAST_M_Sham) +
                                                     (sPWK_M_Cocaine - sPWK_M_Sham) +
                                                     (sWSB_M_Cocaine - sWSB_M_Sham),
                                                   female_cocaine = (s129_F_Cocaine - s129_F_Sham) +
                                                     (sAJ_F_Cocaine - sAJ_F_Sham) +
                                                     (sB6_F_Cocaine - sB6_F_Sham) +
                                                     (sNOD_F_Cocaine - sNOD_F_Sham) +
                                                     (sNZO_F_Cocaine - sNZO_F_Sham) +
                                                     (sCAST_F_Cocaine - sCAST_F_Sham) +
                                                     (sPWK_F_Cocaine - sPWK_F_Sham) +
                                                     (sWSB_F_Cocaine - sWSB_F_Sham),
                                                   male_cocaine = (s129_M_Cocaine - s129_M_Sham) +
                                                     (sAJ_M_Cocaine - sAJ_M_Sham) +
                                                     (sB6_M_Cocaine - sB6_M_Sham) +
                                                     (sNOD_M_Cocaine - sNOD_M_Sham) +
                                                     (sNZO_M_Cocaine - sNZO_M_Sham) +
                                                     (sCAST_M_Cocaine - sCAST_M_Sham) +
                                                     (sPWK_M_Cocaine - sPWK_M_Sham) +
                                                     (sWSB_M_Cocaine - sWSB_M_Sham),
                                                   s129_male = s129_M_Cocaine - s129_M_Sham,
                                                   sAJ_male = sAJ_M_Cocaine - sAJ_M_Sham,
                                                   sB6_male = sB6_M_Cocaine - sB6_M_Sham,
                                                   sNOD_male = sNOD_M_Cocaine - sNOD_M_Sham,
                                                   sNZO_male = sNZO_M_Cocaine - sNZO_M_Sham,
                                                   sCAST_male = sCAST_M_Cocaine - sCAST_M_Sham,
                                                   sPWK_male = sPWK_M_Cocaine - sPWK_M_Sham,
                                                   sWSB_male = sWSB_M_Cocaine - sWSB_M_Sham,
                                                   s129_female = s129_F_Cocaine - s129_F_Sham,
                                                   sAJ_female = sAJ_F_Cocaine - sAJ_F_Sham,
                                                   sB6_female = sB6_F_Cocaine - sB6_F_Sham,
                                                   sNOD_female = sNOD_F_Cocaine - sNOD_F_Sham,
                                                   sNZO_female = sNZO_F_Cocaine - sNZO_F_Sham,
                                                   sCAST_female = sCAST_F_Cocaine - sCAST_F_Sham,
                                                   sPWK_female = sPWK_F_Cocaine - sPWK_F_Sham,
                                                   sWSB_female = sWSB_F_Cocaine - sWSB_F_Sham,
                                                   levels = colnames(modmat_contrasts))


# limma
lmFit_model = lmFit(founders_voom, design = modmat_contrasts, method = "robust", weights = NULL)
lmFit_contrasts = contrasts.fit(lmFit_model, founders_cocaine_overall_contrasts)
lmFit_contrasts = eBayes(lmFit_contrasts, trend = TRUE, robust = TRUE)

# edgeR
# founders_DGEList = estimateGLMRobustDisp(founders_DGEList, design = modmat_edgeR_contrasts)
# founders_DGEList = glmFit(founders_DGEList, design = modmat_edgeR_contrasts)

toptable_cocaine = as.data.frame(topTable(lmFit_contrasts,
                                          coef = "all_cocaine",
                                          number = nrow(founders_voom)))
toptable_cocaine_female = as.data.frame(topTable(lmFit_contrasts,
                                                 coef = "female_cocaine",
                                                 number = nrow(founders_voom)))
toptable_cocaine_male = as.data.frame(topTable(lmFit_contrasts,
                                               coef = "male_cocaine",
                                               number = nrow(founders_voom)))
toptable_s129_male = as.data.frame(topTable(lmFit_contrasts,
                                            coef = "s129_male",
                                            number = nrow(founders_voom)))
toptable_sAJ_male = as.data.frame(topTable(lmFit_contrasts,
                                           coef = "sAJ_male",
                                           number = nrow(founders_voom)))
toptable_sB6_male = as.data.frame(topTable(lmFit_contrasts,
                                           coef = "sB6_male",
                                           number = nrow(founders_voom)))
toptable_sNOD_male = as.data.frame(topTable(lmFit_contrasts,
                                            coef = "sNOD_male",
                                            number = nrow(founders_voom)))
toptable_sNZO_male = as.data.frame(topTable(lmFit_contrasts,
                                            coef = "sNZO_male",
                                            number = nrow(founders_voom)))
toptable_sCAST_male = as.data.frame(topTable(lmFit_contrasts,
                                             coef = "sCAST_male",
                                             number = nrow(founders_voom)))
toptable_sPWK_male = as.data.frame(topTable(lmFit_contrasts,
                                            coef = "sPWK_male",
                                            number = nrow(founders_voom)))
toptable_sWSB_male = as.data.frame(topTable(lmFit_contrasts,
                                            coef = "sWSB_male",
                                            number = nrow(founders_voom)))
toptable_s129_female = as.data.frame(topTable(lmFit_contrasts,
                                              coef = "s129_female",
                                              number = nrow(founders_voom)))
toptable_sAJ_female = as.data.frame(topTable(lmFit_contrasts,
                                             coef = "sAJ_female",
                                             number = nrow(founders_voom)))
toptable_sB6_female = as.data.frame(topTable(lmFit_contrasts,
                                             coef = "sB6_female",
                                             number = nrow(founders_voom)))
toptable_sNOD_female = as.data.frame(topTable(lmFit_contrasts,
                                              coef = "sNOD_female",
                                              number = nrow(founders_voom)))
toptable_sNZO_female = as.data.frame(topTable(lmFit_contrasts,
                                              coef = "sNZO_female",
                                              number = nrow(founders_voom)))
toptable_sCAST_female = as.data.frame(topTable(lmFit_contrasts,
                                               coef = "sCAST_female",
                                               number = nrow(founders_voom)))
toptable_sPWK_female = as.data.frame(topTable(lmFit_contrasts,
                                              coef = "sPWK_female",
                                              number = nrow(founders_voom)))
toptable_sWSB_female = as.data.frame(topTable(lmFit_contrasts,
                                              coef = "sWSB_female",
                                              number = nrow(founders_voom)))
new_toptables = ls()[grep("^toptable",ls())]
```

Getting gene symbols from Ensembl using `biomaRt`.

```{r get_gene_symbols}
maRt = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
               host = "oct2018.archive.ensembl.org",
               dataset = "mmusculus_gene_ensembl")
maRt_filter = "ensembl_gene_id"
maRt_attributes = c("ensembl_gene_id","external_gene_name")
Mm_maRt = getBM(maRt_attributes, maRt_filter, row.names(toptable_cocaine), maRt)
```

Making human:mouse key for MSET

```{r human_to_mouse_key}
Mm_Hs_key = Mm_Hs_homology_1to1 %>%
  dplyr::select(HomoloGene_ID, Hs_Symbol, Mm_Symbol) %>%
  mutate(HomoloGene_ID_chr = paste0("hg_",as.character(HomoloGene_ID))) %>%
  dplyr::select(HomoloGene_ID_chr, Hs_Symbol, Mm_Symbol)
row.names(Mm_Hs_key) = Mm_Hs_key$Mm_Symbol
Mm_maRt$HomoloGene_ID_chr = Mm_Hs_key[Mm_maRt$external_gene_name,"HomoloGene_ID_chr"]
Mm_maRt = Mm_maRt[which(!is.na(Mm_maRt$HomoloGene_ID_chr)),]
row.names(Mm_maRt) = Mm_maRt$ensembl_gene_id
```

Loading human gene set.

```{r}
Hs_brain_cocaine = read.table("~/Box/projects/founders_rnaseq/misc/HUMAN_DEGS_Mouse_orthologs.txt",
                              sep = "\t", header = TRUE, stringsAsFactors = FALSE, quote = "")
row.names(Hs_brain_cocaine) = Hs_brain_cocaine$hgnc_symbol
row.names(Mm_Hs_key) = Mm_Hs_key$Hs_Symbol
Hs_brain_cocaine$HomoloGene_ID_chr = Mm_Hs_key[Hs_brain_cocaine$hgnc_symbol,"HomoloGene_ID_chr"]
Hs_brain_cocaine_genes = unique(Hs_brain_cocaine$HomoloGene_ID_chr[which(Hs_brain_cocaine$HomoloGene_ID_chr %in% Mm_Hs_key$HomoloGene_ID_chr)])
```

Searching for overlaps.

```{r}
strains = unique(founders_strain)
MSET_B = 1e6
MSET_pval_df = matrix(nrow = 8, ncol = 8)
row.names(MSET_pval_df) = strains
colnames(MSET_pval_df) = c("male_q_0p01","male_q_0p05","male_q_0p10","male_q_0p20",
                           "female_q_0p01","female_q_0p05","female_q_0p10","female_q_0p20")
MSET_ngene_df = matrix(nrow = 8, ncol = 8)
row.names(MSET_ngene_df) = strains
colnames(MSET_ngene_df) = c("male_n_100","male_n_250","male_n_500","male_n_1000",
                            "female_n_100","female_n_250","female_n_500","female_n_1000")

for (i in strains) {
  for (j in c("male","female")) {
    toptable_ij = eval(parse(text = paste0("toptable_s",i,"_",j)))
    toptable_ij$qvalue = qvalue(toptable_ij$P.Value)$qvalues
    
    n_100 = Mm_maRt[row.names(toptable_ij)[1:100],"HomoloGene_ID_chr"]
    n_100 = n_100[which(n_100 %in% Mm_maRt$HomoloGene_ID_chr)]
    cat("Strain ",i," ",j,": top 100 genes.\n", sep = "")
    n_100_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                          n_100,
                                          Mm_Hs_key$HomoloGene_ID_chr,
                                          Mm_maRt$HomoloGene_ID_chr,
                                          B = MSET_B)
    cat("MSET overlap p-value is: ",
        round(n_100_MSET$p.value + (1 / MSET_B), 4),
        ".\n",sep="")
    MSET_ngene_df[i,paste0(j,"_n_100")] = n_100_MSET$p.value + (1 / MSET_B)
    
    n_250 = Mm_maRt[row.names(toptable_ij)[1:250],"HomoloGene_ID_chr"]
    n_250 = n_250[which(n_250 %in% Mm_maRt$HomoloGene_ID_chr)]
    cat("Strain ",i," ",j,": top 250 genes.\n", sep = "")
    n_250_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                          n_250,
                                          Mm_Hs_key$HomoloGene_ID_chr,
                                          Mm_maRt$HomoloGene_ID_chr,
                                          B = MSET_B)
    cat("MSET overlap p-value is: ",
        round(n_250_MSET$p.value + (1 / MSET_B), 4),
        ".\n",sep="")
    MSET_ngene_df[i,paste0(j,"_n_250")] = n_250_MSET$p.value + (1 / MSET_B)
    
    n_500 = Mm_maRt[row.names(toptable_ij)[1:500],"HomoloGene_ID_chr"]
    n_500 = n_500[which(n_500 %in% Mm_maRt$HomoloGene_ID_chr)]
    cat("Strain ",i," ",j,": top 500 genes.\n", sep = "")
    n_500_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                          n_500,
                                          Mm_Hs_key$HomoloGene_ID_chr,
                                          Mm_maRt$HomoloGene_ID_chr,
                                          B = MSET_B)
    cat("MSET overlap p-value is: ",
        round(n_500_MSET$p.value + (1 / MSET_B), 4),
        ".\n",sep="")
    MSET_ngene_df[i,paste0(j,"_n_500")] = n_500_MSET$p.value + (1 / MSET_B)
    
    n_1000 = Mm_maRt[row.names(toptable_ij)[1:1000],"HomoloGene_ID_chr"]
    n_1000 = n_1000[which(n_1000 %in% Mm_maRt$HomoloGene_ID_chr)]
    cat("Strain ",i," ",j,": top 1000 genes.\n", sep = "")
    n_1000_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                           n_1000,
                                           Mm_Hs_key$HomoloGene_ID_chr,
                                           Mm_maRt$HomoloGene_ID_chr,
                                           B = MSET_B)
    cat("MSET overlap p-value is: ",
        round(n_1000_MSET$p.value + (1 / MSET_B), 4),
        ".\n",sep="")
    MSET_ngene_df[i,paste0(j,"_n_1000")] = n_1000_MSET$p.value + (1 / MSET_B)
    
    p_0p01 = Mm_maRt[row.names(toptable_ij)[which(toptable_ij$qvalue < 0.01)],"HomoloGene_ID_chr"]
    p_0p01 = p_0p01[which(p_0p01 %in% Mm_maRt$HomoloGene_ID_chr)]
    cat("Strain ",i," ",j,": total of ", length(p_0p01), " genes q < 0.01\n", sep = "")
    p_0p01_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                           p_0p01,
                                           Mm_Hs_key$HomoloGene_ID_chr,
                                           Mm_maRt$HomoloGene_ID_chr,
                                           B = MSET_B)
    cat("MSET overlap p-value is: ",
        round(p_0p01_MSET$p.value + (1 / MSET_B), 4),
        ".\n",sep="")
    MSET_pval_df[i,paste0(j,"_q_0p01")] = p_0p01_MSET$p.value + (1 / MSET_B)
    
    p_0p05 = Mm_maRt[row.names(toptable_ij)[which(toptable_ij$qvalue < 0.05)],"HomoloGene_ID_chr"]
    p_0p05 = p_0p05[which(p_0p05 %in% Mm_maRt$HomoloGene_ID_chr)]
    cat("Strain ",i," ",j,": total of ", length(p_0p05), " genes q < 0.05\n", sep = "")
    p_0p05_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                           p_0p05,
                                           Mm_Hs_key$HomoloGene_ID_chr,
                                           Mm_maRt$HomoloGene_ID_chr,
                                           B = MSET_B)
    cat("MSET overlap p-value is: ",
        round(p_0p05_MSET$p.value + (1 / MSET_B), 4),
        ".\n",sep="")
    MSET_pval_df[i,paste0(j,"_q_0p05")] = p_0p05_MSET$p.value + (1 / MSET_B)
        
    p_0p10 = Mm_maRt[row.names(toptable_ij)[which(toptable_ij$qvalue < 0.10)],"HomoloGene_ID_chr"]
    p_0p10 = p_0p10[which(p_0p10 %in% Mm_maRt$HomoloGene_ID_chr)]
    cat("Strain ",i," ",j,": total of ", length(p_0p10), " genes q < 0.10\n", sep = "")
    p_0p10_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                           p_0p10,
                                           Mm_Hs_key$HomoloGene_ID_chr,
                                           Mm_maRt$HomoloGene_ID_chr,
                                           B = MSET_B)
    cat("MSET overlap p-value is: ",
        round(p_0p10_MSET$p.value + (1 / MSET_B), 4),
        ".\n",sep="")
    MSET_pval_df[i,paste0(j,"_q_0p10")] = p_0p10_MSET$p.value + (1 / MSET_B)
    
    p_0p20 = Mm_maRt[row.names(toptable_ij)[which(toptable_ij$qvalue < 0.20)],"HomoloGene_ID_chr"]
    p_0p20 = p_0p20[which(p_0p20 %in% Mm_maRt$HomoloGene_ID_chr)]
    cat("Strain ",i," ",j,": total of ", length(p_0p20), " genes q < 0.20\n", sep = "")
    p_0p20_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                           p_0p20,
                                           Mm_Hs_key$HomoloGene_ID_chr,
                                           Mm_maRt$HomoloGene_ID_chr,
                                           B = MSET_B)
    cat("MSET overlap p-value is: ",
        round(p_0p20_MSET$p.value + (1 / MSET_B), 4),
        ".\n",sep="")
    MSET_pval_df[i,paste0(j,"_q_0p20")] = p_0p20_MSET$p.value + (1 / MSET_B)
    
    if (exists("top_genes")) {
      top_genes = unique(c(top_genes, row.names(toptable_ij[1:100,])))
    } else {
      top_genes = row.names(toptable_ij[1:100,])
    }
    
  }
}

MSET_ngene_spread_df = as.data.frame(MSET_ngene_df) %>% 
  rownames_to_column() %>% 
  as_tibble() %>% 
  gather(key = threshold, value = mset_pvalue, -rowname) %>%
  mutate(type = gsub("^(\\w{4,6})_n_\\d{3,4}$","\\1",threshold), select_method = "ngene")

MSET_pval_spread_df = as.data.frame(MSET_pval_df) %>% 
  rownames_to_column() %>% 
  as_tibble() %>% 
  gather(key = threshold, value = mset_pvalue, -rowname) %>%
  mutate(type = gsub("^(\\w{4,6})_.*$","\\1",threshold), select_method = "q")
```

Showing number of genes data frame.

```{r}
MSET_ngene_df
```

Showing founders p-value data frame.

```{r}
MSET_pval_df
```

Running MSET on the linear modeling toptables.

```{r}
load("~/Box/projects/founders_rnaseq/founders_RSEM_expression_plots_data_2019-12-19.Rdata")
founder_toptables = ls()[grep("^founders.*toptable$",ls())]
founder_toptables = founder_toptables[grep("cocaine",founder_toptables)]
founder_toptables = c(founder_toptables, "toptable_cocaine","toptable_cocaine_female","toptable_cocaine_male")

founders_pval_df = matrix(nrow = length(founder_toptables), ncol = 4)
row.names(founders_pval_df) = founder_toptables
founders_ngene_df = founders_pval_df
colnames(founders_pval_df) = c("q_0p01","q_0p05","q_0p10","q_0p20")
colnames(founders_ngene_df) = c("n_100","n_250","n_500","n_1000")

for (k in founder_toptables) {
  toptable_k = eval(parse(text = k))
  toptable_k$qvalue = qvalue(toptable_k$P.Value)$qvalues
  
  n_100 = Mm_maRt[row.names(toptable_k)[1:100],"HomoloGene_ID_chr"]
  n_100 = n_100[which(n_100 %in% Mm_maRt$HomoloGene_ID_chr)]
  cat("Toptable ", k, ": top 100 genes.\n", sep = "")
  n_100_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                        n_100,
                                        Mm_Hs_key$HomoloGene_ID_chr,
                                        Mm_maRt$HomoloGene_ID_chr,
                                        B = MSET_B)
  cat("MSET overlap p-value is: ",
      round(n_100_MSET$p.value + (1 / MSET_B), 4),
      ".\n",sep="")
  founders_ngene_df[k,"n_100"] = n_100_MSET$p.value + (1 / MSET_B)
  
  n_250 = Mm_maRt[row.names(toptable_k)[1:250],"HomoloGene_ID_chr"]
  n_250 = n_250[which(n_250 %in% Mm_maRt$HomoloGene_ID_chr)]
  cat("Toptable ", k, ": top 250 genes.\n", sep = "")
  n_250_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                        n_250,
                                        Mm_Hs_key$HomoloGene_ID_chr,
                                        Mm_maRt$HomoloGene_ID_chr,
                                        B = MSET_B)
  cat("MSET overlap p-value is: ",
      round(n_250_MSET$p.value + (1 / MSET_B), 4),
      ".\n",sep="")
  founders_ngene_df[k,"n_250"] = n_250_MSET$p.value + (1 / MSET_B)
  
  n_500 = Mm_maRt[row.names(toptable_k)[1:500],"HomoloGene_ID_chr"]
  n_500 = n_500[which(n_500 %in% Mm_maRt$HomoloGene_ID_chr)]
  cat("Toptable ", k, ": top 500 genes.\n", sep = "")
  n_500_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                        n_500,
                                        Mm_Hs_key$HomoloGene_ID_chr,
                                        Mm_maRt$HomoloGene_ID_chr,
                                        B = MSET_B)
  cat("MSET overlap p-value is: ",
      round(n_500_MSET$p.value + (1 / MSET_B), 4),
      ".\n",sep="")
  founders_ngene_df[k,"n_500"] = n_500_MSET$p.value + (1 / MSET_B)
  
  n_1000 = Mm_maRt[row.names(toptable_k)[1:1000],"HomoloGene_ID_chr"]
  n_1000 = n_1000[which(n_1000 %in% Mm_maRt$HomoloGene_ID_chr)]
  cat("Toptable ", k, ": top 1000 genes.\n", sep = "")
  n_1000_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                        n_1000,
                                        Mm_Hs_key$HomoloGene_ID_chr,
                                        Mm_maRt$HomoloGene_ID_chr,
                                        B = MSET_B)
  cat("MSET overlap p-value is: ",
      round(n_1000_MSET$p.value + (1 / MSET_B), 4),
      ".\n",sep="")
  founders_ngene_df[k,"n_1000"] = n_1000_MSET$p.value + (1 / MSET_B)
  
  p_0p01 = Mm_maRt[row.names(toptable_k)[which(toptable_k$qvalue < 0.01)],"HomoloGene_ID_chr"]
  p_0p01 = p_0p01[which(p_0p01 %in% Mm_maRt$HomoloGene_ID_chr)]
  cat("Toptable ",k, ": total of ", length(p_0p01), " genes q < 0.01\n", sep = "")
  p_0p01_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                         p_0p01,
                                         Mm_Hs_key$HomoloGene_ID_chr,
                                         Mm_maRt$HomoloGene_ID_chr,
                                         B = MSET_B)
  cat("MSET overlap p-value is: ",
      round(p_0p01_MSET$p.value + (1 / MSET_B), 4),
      ".\n",sep="")
  founders_pval_df[k,"q_0p01"] = p_0p01_MSET$p.value + (1 / MSET_B)
  
  p_0p05 = Mm_maRt[row.names(toptable_k)[which(toptable_k$qvalue < 0.05)],"HomoloGene_ID_chr"]
  p_0p05 = p_0p05[which(p_0p05 %in% Mm_maRt$HomoloGene_ID_chr)]
  cat("Toptable ",k, ": total of ", length(p_0p05), " genes q < 0.05\n", sep = "")
  p_0p05_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                         p_0p05,
                                         Mm_Hs_key$HomoloGene_ID_chr,
                                         Mm_maRt$HomoloGene_ID_chr,
                                         B = MSET_B)
  cat("MSET overlap p-value is: ",
      round(p_0p05_MSET$p.value + (1 / MSET_B), 4),
      ".\n",sep="")
  founders_pval_df[k,"q_0p05"] = p_0p05_MSET$p.value + (1 / MSET_B)
  
  p_0p10 = Mm_maRt[row.names(toptable_k)[which(toptable_k$qvalue < 0.10)],"HomoloGene_ID_chr"]
  p_0p10 = p_0p10[which(p_0p10 %in% Mm_maRt$HomoloGene_ID_chr)]
  cat("Toptable ",k, ": total of ", length(p_0p10), " genes q < 0.10\n", sep = "")
  p_0p10_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                         p_0p10,
                                         Mm_Hs_key$HomoloGene_ID_chr,
                                         Mm_maRt$HomoloGene_ID_chr,
                                         B = MSET_B)
  cat("MSET overlap p-value is: ",
      round(p_0p10_MSET$p.value + (1 / MSET_B), 4),
      ".\n",sep="")
  founders_pval_df[k,"q_0p10"] = p_0p10_MSET$p.value + (1 / MSET_B)
  
  p_0p20 = Mm_maRt[row.names(toptable_k)[which(toptable_k$qvalue < 0.20)],"HomoloGene_ID_chr"]
  p_0p20 = p_0p20[which(p_0p20 %in% Mm_maRt$HomoloGene_ID_chr)]
  cat("Toptable ",k, ": total of ", length(p_0p20), " genes q < 0.20\n", sep = "")
  p_0p20_MSET = msaul::MSET.overlap.test(Hs_brain_cocaine_genes,
                                         p_0p20,
                                         Mm_Hs_key$HomoloGene_ID_chr,
                                         Mm_maRt$HomoloGene_ID_chr,
                                         B = MSET_B)
  cat("MSET overlap p-value is: ",
      round(p_0p20_MSET$p.value + (1 / MSET_B), 4),
      ".\n",sep="")
  founders_pval_df[k,"q_0p20"] = p_0p20_MSET$p.value + (1 / MSET_B)
  
}

founders_ngene_spread_df = as.data.frame(founders_ngene_df) %>% 
  rownames_to_column() %>% 
  as_tibble() %>% 
  gather(key = threshold, value = mset_pvalue, -rowname) %>%
  mutate(type = "founders", select_method = "ngene")

founders_pval_spread_df = as.data.frame(founders_pval_df) %>% 
  rownames_to_column() %>% 
  as_tibble() %>% 
  gather(key = threshold, value = mset_pvalue, -rowname) %>%
  mutate(type = "founders", select_method = "q")
```

Showing founders number of genes data frame.

```{r}
founders_ngene_df
```

Showing founders p-value data frame.

```{r}
founders_pval_df
```

### Plotting MSET results

```{r}
founders_MSET_results = rbind(MSET_pval_spread_df,
                              MSET_ngene_spread_df,
                              founders_pval_spread_df,
                              founders_ngene_spread_df)
founders_MSET_results$mset_pvalue = ifelse(founders_MSET_results$mset_pvalue > 1, 1,
                                           founders_MSET_results$mset_pvalue)
founders_MSET_results$type = gsub("_q$","",founders_MSET_results$type)

colnames(founders_MSET_results)[1] = "geneset"
founders_MSET_results$threshold = gsub("^male_","",founders_MSET_results$threshold)
founders_MSET_results$threshold = gsub("^female_","",founders_MSET_results$threshold)
founders_MSET_results$highlight = ifelse((founders_MSET_results$type == "male" & founders_MSET_results$geneset == "WSB") |
                                         (founders_MSET_results$type == "female" & founders_MSET_results$geneset == "NOD"),
                                         "YES", "no")

founders_MSET_results$threshold = factor(founders_MSET_results$threshold,
                                         levels = c("q_0p01","q_0p05","q_0p10","q_0p20",
                                                    "n_100","n_250","n_500","n_1000"),
                                         ordered = TRUE)
levels(founders_MSET_results$threshold) = c("q < 0.01","q < 0.05", "q < 0.10", "q < 0.20",
                                            "Top 100", "Top 250", "Top 500", "Top 1000")

founders_MSET_results$geneset = factor(founders_MSET_results$geneset,
                                       levels = c(rev(strain_key$collaborative_cross_abbreviation),
                                                  "founders_cocaine_toptable","founders_strain_cocaine_toptable",
                                                  "founders_sex_cocaine_toptable","founders_strain_sex_cocaine_toptable",
                                                  "toptable_cocaine","toptable_cocaine_male","toptable_cocaine_female"),
                                       ordered = TRUE)
levels(founders_MSET_results$geneset) = c(rev(strain_key$strain),
                                          "ANOVA: Cocaine",
                                          "ANOVA: Strain x Cocaine",
                                          "ANOVA: Sex x Cocaine",
                                          "ANOVA: Strain x Sex x Cocaine",
                                          "Contrasts: Both Sexes Cocaine vs Saline",
                                          "Contrasts: Males Cocaine vs Saline",
                                          "Contrasts: Females Cocaine vs Saline")

founders_MSET_results$type = factor(founders_MSET_results$type,
                                    levels = c("female","male","founders"),
                                    ordered = TRUE)
levels(founders_MSET_results$type) = c("Female Cocaine vs Saline",
                                       "Male Cocaine vs Saline",
                                       "Omnibus and Multi-Strain Methods")
founders_MSET_results$select_method = factor(founders_MSET_results$select_method,
                                             levels = c("q","ngene"),
                                             ordered = TRUE)
levels(founders_MSET_results$select_method) = c("Significance (q)",
                                                "Number of Genes")

founders_MSET_results_plot = ggplot(data = as.data.frame(founders_MSET_results), 
                                    aes(x = threshold, y = geneset, fill = mset_pvalue)) +
  geom_tile(size = 0.1, color = "#FFFFFF") +
  facet_grid(type ~ select_method, scales = "free", space = "free") +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        panel.grid = element_blank()) +
  xlab(NULL) + ylab(NULL) + ggtitle("MSET Results: Mouse vs Human") +
  scale_fill_viridis(option = "C", name = "Overlap P-Value")

founders_MSET_results_plot
```

Plotting just the individual strain q correction values

```{r strain_significance_results}
strain_significance_results = as.data.frame(founders_MSET_results[which(founders_MSET_results$type != "Omnibus and Multi-Strain Methods" & founders_MSET_results$select_method == "Significance (q)"),])
strain_significance_results = strain_significance_results[order(strain_significance_results$mset_pvalue),]
strain_significance_results$qvalue = qvalue(strain_significance_results$mset_pvalue)$qvalues
strain_significance_results$significant = strain_significance_results$qvalue < 0.05
strain_significance_results$sig_label = ifelse(strain_significance_results$qvalue < 0.20,".","")
strain_significance_results[which(strain_significance_results$qvalue < 0.05),"sig_label"] = "*"
                                    
strain_MSET_results_plot = ggplot(data = strain_significance_results,
                                  aes(x = threshold, y = geneset, fill = mset_pvalue)) +
  geom_tile(size = 0.1, color = "#FFFFFF") +
  geom_text(aes(label = sig_label), color = "#FFFFFF", size = 12) +
  facet_grid(. ~ type, scales = "free", space = "free") +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        panel.grid = element_blank()) +
  xlab(NULL) + ylab(NULL) + ggtitle("Mouse vs Human Overlap Results") +
  scale_fill_viridis(option = "A", name = "Overlap P-Value") +
  theme(legend.position = "bottom")

strain_MSET_results_plot
```

### Plotting interesting relationships: NOD females and WSB males

Looking at fold-change for WSB male findings at q < 0.05 and NOD female findings at q < 0.01.

```{r}
# Starting by getting WSB DEGs at q < 0.05
toptable_sWSB_male$qvalue = qvalue(toptable_sWSB_male$P.Value)$qvalues
WSB_M_DEGs = Mm_maRt[row.names(toptable_sWSB_male)[which(toptable_sWSB_male$qvalue < 0.05)],"HomoloGene_ID_chr"]
WSB_M_DEGs = WSB_M_DEGs[which(WSB_M_DEGs %in% Mm_Hs_key$HomoloGene_ID_chr)]
Mm_Hs_key$is_WSB_M_DEG = Mm_Hs_key$HomoloGene_ID_chr %in% WSB_M_DEGs

# Getting NOD DEGs at q < 0.01
toptable_sNOD_female$qvalue = qvalue(toptable_sNOD_female$P.Value)$qvalues
NOD_F_DEGs = Mm_maRt[row.names(toptable_sNOD_female)[which(toptable_sNOD_female$qvalue < 0.01)],"HomoloGene_ID_chr"]
NOD_F_DEGs = NOD_F_DEGs[which(NOD_F_DEGs %in% Mm_Hs_key$HomoloGene_ID_chr)]
Mm_Hs_key$is_NOD_F_DEG = Mm_Hs_key$HomoloGene_ID_chr %in% NOD_F_DEGs

# Getting human cocaine-related genes
Mm_Hs_key$is_Hs_cocaine = Mm_Hs_key$HomoloGene_ID_chr %in% Hs_brain_cocaine_genes

# Making data frames to hold plotting data
Mm_WSB_M_Hs = Mm_Hs_key[which(Mm_Hs_key$is_Hs_cocaine & Mm_Hs_key$is_WSB_M_DEG),]
Mm_NOD_F_Hs = Mm_Hs_key[which(Mm_Hs_key$is_Hs_cocaine & Mm_Hs_key$is_NOD_F_DEG),]

# Getting WSB Male fold-change
Mm_WSB_M_Hs$Strain = rep("WSB/EiJ (q < 0.05)", times = nrow(Mm_WSB_M_Hs))
Mm_WSB_M_Hs$Sex = rep("Male", times = nrow(Mm_WSB_M_Hs))
Mm_WSB_M_Hs$human_fc = Hs_brain_cocaine[Mm_WSB_M_Hs$Hs_Symbol,"log2FoldChange"]
Mm_WSB_M_Mm_IDs = Mm_WSB_M_Hs$HomoloGene_ID_chr
Mm_WSB_M_Mm_IDs = Mm_maRt[which(Mm_maRt$HomoloGene_ID_chr %in% Mm_WSB_M_Mm_IDs),]
Mm_WSB_M_Hs$mouse_fc = toptable_sWSB_male[Mm_WSB_M_Mm_IDs$ensembl_gene_id,"logFC"]

# Getting NOD Female fold-change
Mm_NOD_F_Hs$Strain = rep("NOD/ShiLtJ (q < 0.01)", times = nrow(Mm_NOD_F_Hs))
Mm_NOD_F_Hs$Sex = rep("Female", times = nrow(Mm_NOD_F_Hs))
Mm_NOD_F_Hs$human_fc = Hs_brain_cocaine[Mm_NOD_F_Hs$Hs_Symbol,"log2FoldChange"]
Mm_NOD_F_Mm_IDs = Mm_NOD_F_Hs$HomoloGene_ID_chr
Mm_NOD_F_Mm_IDs = Mm_maRt[which(Mm_maRt$HomoloGene_ID_chr %in% Mm_NOD_F_Mm_IDs),]
Mm_NOD_F_Hs$mouse_fc = toptable_sNOD_female[Mm_NOD_F_Mm_IDs$ensembl_gene_id,"logFC"]

# Making table for plotting
Mm_Hs_FC_df = rbind(Mm_NOD_F_Hs, Mm_WSB_M_Hs)

# Highlighting specific genes
genes_to_highlight = c("Arc","Junb","Ndufa5","Ndufaf2","Cox6c2")
Mm_Hs_FC_df$highlight = ifelse(Mm_Hs_FC_df$Mm_Symbol %in% genes_to_highlight,
                               Mm_Hs_FC_df$Mm_Symbol, "")

# Plotting fold-change agreement
Mm_Hs_figure = ggplot(data = Mm_Hs_FC_df, aes(x = human_fc, 
                                              y = mouse_fc, 
                                              color = Strain, 
                                              shape = Sex,
                                              label = highlight)) +
  geom_point(size = 2.5) +
  geom_label_repel(show.legend = FALSE) +
  theme_classic() +
  geom_hline(aes(yintercept = 0), linetype = "dashed", color = "#AAAAAA") +
  geom_vline(aes(xintercept = 0), linetype = "dashed", color = "#AAAAAA") +
  scale_color_manual(values = c("#0064C9","#B10DC9")) +
  scale_shape_manual(values = c(16,15)) +
  xlab("Human log2(fold-change)") +
  ylab("Mouse log2(fold-change)") +
  theme(legend.position = "bottom")

Mm_Hs_figure
```

### Saving Data

Saving all data

```{r}
save(list = ls(), file = paste("founders_RSEM_overlap_data_",format(Sys.time(),"%Y-%m-%d"),".Rdata",sep=""))
```

## Reproducibility Information

### R session and OS information

```{r sessionInfo}
founders_RSEM_overlap_data_sessionInfo = sessionInfo()
founders_RSEM_overlap_data_sessionInfo
saveRDS(founders_RSEM_overlap_data_sessionInfo,paste("./founders_RSEM_overlap_data_sessionInfo_",format(Sys.time(),"%Y-%m-%d"),".RDS",sep=""))
```

### Document Control

This document was prepared using RMarkdown in RStudio.